lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,748 @@
+	}
+
+	fn send_tx_add_output(mut self, serial_id: SerialId, output: TxOut) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.outputs.insert(serial_id, output);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_remove_input(mut self, serial_id: SerialId) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.inputs.remove(&serial_id);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_remove_output(mut self, serial_id: SerialId) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.outputs.remove(&serial_id);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_abort(mut self) -> InteractiveTxStateMachine<NegotiationAborted> {
+		// A sending node:
+		// 	- MUST NOT have already transmitted tx_signatures
+		// 	- SHOULD forget the current negotiation and reset their state.
+		todo!();
+	}
+
+	fn receive_tx_abort(mut self) -> InteractiveTxStateMachine<NegotiationAborted> {
+		todo!();
+	}
+
+	fn is_current_transaction_state_able_to_complete(&self) -> Result<(), AbortReason> {
+		let tx_to_validate = Transaction {
+			version: self.context.base_tx.version,
+			lock_time: self.context.base_tx.lock_time,
+			input: self.context.inputs.values().map(|p| p.input.clone()).collect(),
+			output: self.context.outputs.values().cloned().collect(),
+		};
+
+		// The receiving node:
+		// MUST fail the negotiation if:
+
+		// - the peer's total input satoshis is less than their outputs
+		let total_input_amount: u64 = self.context.inputs.values().map(|p| p.prev_output.value).sum();
+		let total_output_amount = tx_to_validate.output.iter().map(|output| output.value).sum();
+		if total_input_amount < total_output_amount {
+			return Err(AbortReason::OutputsExceedInputs);
+		}
+
+		// - there are more than 252 inputs
+		// - there are more than 252 outputs
+		if self.context.inputs.len() > MAX_INPUTS_OUTPUTS_COUNT ||
+			self.context.outputs.len() > MAX_INPUTS_OUTPUTS_COUNT {
+			return Err(AbortReason::ExceededNumberOfInputsOrOutputs)
+		}
+
+		if tx_to_validate.weight() as u32 > MAX_STANDARD_TX_WEIGHT {
+			return Err(AbortReason::TransactionTooLarge)
+		}
+
+		// TODO:
+		// - Use existing rust-lightning/rust-bitcoin constants.
+		// - How do we enforce their fees cover the witness without knowing its expected length?
+		// 	 - Read eclair's code to see if they do this?
+		const INPUT_WEIGHT: u64 = (32 + 4 + 4) * WITNESS_SCALE_FACTOR as u64;
+		const OUTPUT_WEIGHT: u64 = 8 * WITNESS_SCALE_FACTOR as u64;
+
+		// - the peer's paid feerate does not meet or exceed the agreed feerate (based on the minimum fee).
+		if self.context.holder_is_initiator {
+			let non_initiator_fees_contributed: u64 = self.context.non_initiator_outputs_contributed().map(|output| output.value).sum::<u64>() -
+				self.context.non_initiator_inputs_contributed().map(|input| input.prev_output.value).sum::<u64>();
```

Comment:
This substraction can result in 'attempt to subtract with overflow'.

Reply:
We check that the sum of input amounts are greater than sum of output amounts above and fail early, but for peace of mind we can do saturating subtraction of inputs minus outputs (as you noted below) here anyway.

Also, I believe we need to include the contribution to the funding output amount by the non-initiator here otherwise we'd be counting that contribution towards fees.

So something like (modulo variable names):
`non_initiator_inputs - non_initiator_outputs_sum - non_initiator_amount_contributed`

Of course, the initiator is the one who pays for the weight contributed by the funding output in all cases, but we still need to consider amounts contributed by each party in these calculations.